mesonbuild/options.py

```diff
@@ -192,6 +195,25 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f'OptionKey({self.name!r}, {self.subproject!r}, {self.machine!r})'
 
+    @classmethod
+    def factory(cls, name: str, subproject: str = '', machine: MachineChoice = MachineChoice.HOST) -> OptionKey:
+        """A cached initializer for OptionKeys
+
+        :param name: The name of the option
+        :param subproject: the subproject the option belongs to, defaults to ''
+        :param machine: the machine the subproject belongs to, defaults to MachineChoice.HOST
+        :return: An OptionKey
+        """
+        # This doesn't use lru_cache so that we can re-use the calculated hash
+        # in a cache miss situation.
+        _hash = hash((name, subproject, machine))
+        try:
+            return _OPTIONS_CACHE[_hash]
```

Comment:
It is unlikely, but two objects can have the same hash but not be equal. I'm not sure that the cache lookup based on the hash only is a good idea. If this is done so many times that computing the hash becomes the bottleneck, maybe it is better to simply cache the has.